src/components/mc/cuda/mc_cuda.c

```diff
@@ -234,23 +293,108 @@ static ucc_status_t ucc_mc_cuda_mem_alloc(void **ptr, size_t size)
                  size, st, cudaGetErrorString(st));
         return UCC_ERR_NO_MEMORY;
     }
+    mc_debug(
+        &ucc_mc_cuda.super, "ucc_mc_cuda_mem_alloc size:%zd",
+        size); // TODO: should i keep this? add it also to chunk alloc? in cpu?
+    h->from_pool = 0;
+    *ptr         = h;
+    return UCC_OK;
+}
+
+static ucc_status_t ucc_mc_cuda_chunk_alloc(ucc_mpool_t *mp, size_t *size_p,
+                                            void **chunk_p)
+{
+    *chunk_p = ucc_malloc(*size_p, "mc cuda");
+    if (!*chunk_p) {
+        mc_error(&ucc_mc_cuda.super, "failed to allocate %zd bytes", *size_p);
+        return UCC_ERR_NO_MEMORY;
+    }
 
-    mc_debug(&ucc_mc_cuda.super, "ucc_mc_cuda_mem_alloc size:%ld", size);
     return UCC_OK;
 }
 
-static ucc_status_t ucc_mc_cuda_mem_free(void *ptr)
+static void ucc_mc_cuda_chunk_init(ucc_mpool_t *mp, void *obj, void *chunk)
+{
+    ucc_mc_buffer_header_t *h = (ucc_mc_buffer_header_t *)obj;
+    cudaError_t st = cudaMalloc(&h->addr, MC_CUDA_CONFIG->cuda_elem_size);
+    if (st != cudaSuccess) {
+        cudaGetLastError();
+        mc_error(&ucc_mc_cuda.super,
+                 "failed to allocate %zd bytes, "
+                 "cuda error %d(%s)",
+                 MC_CUDA_CONFIG->cuda_elem_size, st, cudaGetErrorString(st));
```

Comment:
since this function can't return error codes we need a better way to handle failure. maybe reset mc_alloc to "slow_path_only" here as well ?